plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcMetadata.java

```diff
@@ -171,6 +175,33 @@ private JdbcTableHandle flushAttributesAsQuery(ConnectorSession session, JdbcTab
                 handle.getNextSyntheticColumnId());
     }
 
+    private JdbcTableHandle flushTopNAttributesAsQuery(ConnectorSession session, JdbcTableHandle handle, List<SortItem> sortItems, long limit)
+    {
+        verify(!sortItems.isEmpty(), "sortItems is empty");
+        return transformQueryRelationPreparedQuery(session, handle, preparedQuery ->
+                jdbcClient.prepareTopNQuery(preparedQuery, sortItems, limit));
+    }
+
+    private JdbcTableHandle transformQueryRelationPreparedQuery(ConnectorSession session, JdbcTableHandle handle, Function<PreparedQuery, PreparedQuery> transform)
+    {
+        List<JdbcColumnHandle> columns = jdbcClient.getColumns(session, handle);
+
+        if (!(handle.getRelationHandle() instanceof JdbcQueryRelationHandle)) {
+            handle = flushAttributesAsQuery(session, handle);
```

Comment:
Thats not correct (to weak) if the table already has LIMIT or constraint. Current code will produce incorrect results.
Please add tests for that.